src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -389,6 +389,18 @@ protected virtual IEnumerable<SyntaxNode> ExtractNodesIfInHeader(SyntaxNode root
             {
                 yield return localDeclaration;
             }
+
+            // Header: `if(...)`{ };
+            if (syntaxFacts.IsOnIfStatementHeader(root, location, out var ifStatement))
+            {
+                yield return ifStatement;
+            }
+
+            // Header: `foreach (var a in b)` { }
+            if (syntaxFacts.IsOnForeachHeader(root, location, out var foreachStatement))
```

Comment:
Curious where do you handle inside for statement header?

Reply:
We don't handle header of for specially. The idea behind that is that for's expression (declaration, condition, ...) can be way more complex expressions and so the header for "for" is essentially `for(`...`;`...`;`...`)`{...}.

That said, when you brought it up like this it might not be the way we want to do it.

VB's case of extraction the whole block from a header statement is handled (so selecting just the header works fine).